File: java/binary_tree_preorder_traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public List<Integer> preorderTraversal(TreeNode root) {
        List<Integer> res=new ArrayList<>();
        if (root==null) return res;
        Stack<TreeNode> stack=new Stack<>();
        stack.push(root);
        
        while (!stack.empty()) {
            TreeNode t=stack.pop();
            res.add(t.val);
            
            if (t.right!=null) {
                stack.push(t.right);
            }
            if (t.left!=null) {
                stack.push(t.left);
            }
        }
        return res;
    }
}

File: python/binary-tree-preorder-traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def preorderTraversal(self, root: TreeNode) -> List[int]:
        if not root: return []
        res, stk=[], [root]
        while stk:
            t=stk.pop()
            res.append(t.val)
            if t.right: stk.append(t.right)
            if t.left: stk.append(t.left)
        return res

